contribute/design/1.x/migrate/spc-to-cspc-migration.md

```diff
@@ -0,0 +1,190 @@
+* [Proposal](#proposal)
+    * [Proposed Implementation](#proposed-implementation)
+    * [High Level Design](#high-level-design)
+* [Infrastructure Needed](#infrastructure-needed)
+
+## Summary
+
+This design is aimed at providing a design for migrating SPC to CSPC
+via kubernetes job which will take SPC-name as input.
+
+This proposed design will be rolled out in phases. At a high level the design is
+implemented in the following phases:
+- Phase 1: Ability to perform migration from SPC to CSPC using
+  a Kubernetes Job.
+- Phase 2: Allow for migration of old CStor volumes to CSI CStor volumes.
+
+## Motivation
+
+### Goals
+
+- Ease the process of migrating SPC to CSPC by automating it via kubernetes job.
+- Enable day 2 operations on CStor Volumes by migrating them to CSI volumes.
+
+## Proposal
+
+### Proposed Implementation
+
+This design proposes the following key changes while migrating from a SPC to CSPC:
+
+  1. New CSPC CR will be created for the migrating SPC.
+
+  2. New CSPI CRs will be created which will replace the CSP for given SPC.
+
+  3. The BDC with SPC label and finalizer will be replaced by CSPC label and finalizer.
+
+  4. The imported pools in each CSPI will be renamed to cstor-${CSPC uid}.
+
+  5. The CVR with CSP labels and annotations will be replaced by CSPI labels and annotations.
+
+  6. Clean up old SPC and csp after successfull creation of CSPC and cspi objects.
+
+For migrating non-csi volumes to csi volumes following changes are proposed:
+
+  1. New StorageClass with CSPC will be created and this will replace the old strageclass in the PVC of CStorVolume.
+
+  2. Set the PV to `Retain` reclaim policy to prevent deletion of OpenEBS resources.
+
+  3. Recreate old PV with csi spec translation.
+
+  4. Recreate old PVC with updated csi spec PV and StorageClass.
+
+  5. Create CVC CR for the volume with the CV details.
+
+  6. Update ownerreferences of CV, service with CVC and CVRs with CV.
+
+### High Level Design
+
+#### Phase 1: Migration of SPC to CSPC
+
+The migration of SPC will be performed via a job which takes SPC name as one of its argument.
+
+The CSPC CR will be with `reconcile.openebs.io/dependants` annotation to disable reconciliation of CSPI. Once all CSPI are successfully created the annotation will be removed.
+
+Sequentially one CSPI is taken and the corresponding CSP is found using `kubernetes/hostname` label. The CSP deployment is scaled down to avoid multiple pods trying to import the same pool. Next the all the BDC for given CSPI are updated with CSPC information. The CSPI will be patched with the annotation `cstorpoolinstance.openebs.io/oldname`. Then CSPI reconciliation will be enabled which will create the CSPI deployment which will rename and import the pool.
+
+The `cstorpoolinstance.openebs.io/oldname` annotation will be used to rename the pool which was named as `cstor-cspuid` to `cstor-cspcuid`. This annotation will be removed after successful import of the pool.
```

Comment:
It is good to outline — how rename is going to work. For example — is there any `zfs command` involved in doing this. This can help to bring out/highlight risks/corner cases, if any.